// tsconfig-paths/src/register.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const match_path_sync_1 = require("./match-path-sync");
const config_loader_1 = require("./config-loader");
const noOp = () => void 0;
function getCoreModules(builtinModules) {
    builtinModules = builtinModules || [
        "assert",
        "buffer",
        "child_process",
        "cluster",
        "crypto",
        "dgram",
        "dns",
        "domain",
        "events",
        "fs",
        "http",
        "https",
        "net",
        "os",
        "path",
        "punycode",
        "querystring",
        "readline",
        "stream",
        "string_decoder",
        "tls",
        "tty",
        "url",
        "util",
        "v8",
        "vm",
        "zlib",
    ];
    const coreModules = {};
    for (let module of builtinModules) {
        coreModules[module] = true;
    }
    return coreModules;
}
/**
 * Installs a custom module load function that can adhere to paths in tsconfig.
 * Returns a function to undo paths registration.
 */
function register(params) {
    let cwd;
    let explicitParams;
    if (params) {
        cwd = params.cwd;
        if (params.baseUrl || params.paths) {
            explicitParams = params;
        }
    }
    else {
        // eslint-disable-next-line
        const minimist = require("minimist");
        const argv = minimist(process.argv.slice(2), {
            // eslint-disable-next-line id-denylist
            string: ["project"],
            alias: {
                project: ["P"],
            },
        });
        cwd = argv.project;
    }
    const configLoaderResult = (0, config_loader_1.configLoader)({
        cwd: cwd !== null && cwd !== void 0 ? cwd : process.cwd(),
        explicitParams,
    });
    if (configLoaderResult.resultType === "failed") {
        console.warn(`${configLoaderResult.message}. tsconfig-paths will be skipped`);
        return noOp;
    }
    const matchPath = (0, match_path_sync_1.createMatchPath)(configLoaderResult.absoluteBaseUrl, configLoaderResult.paths, configLoaderResult.mainFields, configLoaderResult.addMatchAll);
    // Patch node's module loading
    // eslint-disable-next-line @typescript-eslint/no-require-imports,@typescript-eslint/no-var-requires
    const Module = require("module");
    // eslint-disable-next-line no-underscore-dangle
    const originalResolveFilename = Module._resolveFilename;
    const coreModules = getCoreModules(Module.builtinModules);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any,no-underscore-dangle
    Module._resolveFilename = function (request, _parent) {
        const isCoreModule = coreModules.hasOwnProperty(request);
        if (!isCoreModule) {
            const found = matchPath(request);
            if (found) {
                const modifiedArguments = [found, ...[].slice.call(arguments, 1)]; // Passes all arguments. Even those that is not specified above.
                return originalResolveFilename.apply(this, modifiedArguments);
            }
        }
        return originalResolveFilename.apply(this, arguments);
    };
    return () => {
        // Return node's module loading to original state.
        // eslint-disable-next-line no-underscore-dangle
        Module._resolveFilename = originalResolveFilename;
    };
}
exports.register = register;
