// tsconfig-paths/src/tsconfig-loader.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadTsconfig = exports.walkForTsConfig = exports.tsConfigLoader = void 0;
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
// eslint-disable-next-line @typescript-eslint/no-require-imports
const JSON5 = require("json5");
// eslint-disable-next-line @typescript-eslint/no-require-imports
const StripBom = require("strip-bom");
function tsConfigLoader({ getEnv, cwd, loadSync = loadSyncDefault, }) {
    const TS_NODE_PROJECT = getEnv("TS_NODE_PROJECT");
    const TS_NODE_BASEURL = getEnv("TS_NODE_BASEURL");
    // tsconfig.loadSync handles if TS_NODE_PROJECT is a file or directory
    // and also overrides baseURL if TS_NODE_BASEURL is available.
    const loadResult = loadSync(cwd, TS_NODE_PROJECT, TS_NODE_BASEURL);
    return loadResult;
}
exports.tsConfigLoader = tsConfigLoader;
function loadSyncDefault(cwd, filename, baseUrl) {
    // Tsconfig.loadSync uses path.resolve. This is why we can use an absolute path as filename
    const configPath = resolveConfigPath(cwd, filename);
    if (!configPath) {
        return {
            tsConfigPath: undefined,
            baseUrl: undefined,
            paths: undefined,
        };
    }
    const config = loadTsconfig(configPath);
    return {
        tsConfigPath: configPath,
        baseUrl: baseUrl ||
            (config && config.compilerOptions && config.compilerOptions.baseUrl),
        paths: config && config.compilerOptions && config.compilerOptions.paths,
    };
}
function resolveConfigPath(cwd, filename) {
    if (filename) {
        const absolutePath = fs.lstatSync(filename).isDirectory()
            ? path.resolve(filename, "./tsconfig.json")
            : path.resolve(cwd, filename);
        return absolutePath;
    }
    if (fs.statSync(cwd).isFile()) {
        return path.resolve(cwd);
    }
    const configAbsolutePath = walkForTsConfig(cwd);
    return configAbsolutePath ? path.resolve(configAbsolutePath) : undefined;
}
function walkForTsConfig(directory, readdirSync = fs.readdirSync) {
    const files = readdirSync(directory);
    const filesToCheck = ["tsconfig.json", "jsconfig.json"];
    for (const fileToCheck of filesToCheck) {
        if (files.indexOf(fileToCheck) !== -1) {
            return path.join(directory, fileToCheck);
        }
    }
    const parentDirectory = path.dirname(directory);
    // If we reached the top
    if (directory === parentDirectory) {
        return undefined;
    }
    return walkForTsConfig(parentDirectory, readdirSync);
}
exports.walkForTsConfig = walkForTsConfig;
function loadTsconfig(configFilePath, 
// eslint-disable-next-line no-shadow
existsSync = fs.existsSync, readFileSync = (filename) => fs.readFileSync(filename, "utf8")) {
    if (!existsSync(configFilePath)) {
        return undefined;
    }
    const configString = readFileSync(configFilePath);
    const cleanedJson = StripBom(configString);
    let config;
    try {
        config = JSON5.parse(cleanedJson);
    }
    catch (e) {
        throw new Error(`${configFilePath} is malformed ${e.message}`);
    }
    let extendedConfig = config.extends;
    if (extendedConfig) {
        let base;
        if (Array.isArray(extendedConfig)) {
            base = extendedConfig.reduce((currBase, extendedConfigElement) => mergeTsconfigs(currBase, loadTsconfigFromExtends(configFilePath, extendedConfigElement, existsSync, readFileSync)), {});
        }
        else {
            base = loadTsconfigFromExtends(configFilePath, extendedConfig, existsSync, readFileSync);
        }
        return mergeTsconfigs(base, config);
    }
    return config;
}
exports.loadTsconfig = loadTsconfig;
/**
 * Intended to be called only from loadTsconfig.
 * Parameters don't have defaults because they should use the same as loadTsconfig.
 */
function loadTsconfigFromExtends(configFilePath, extendedConfigValue, 
// eslint-disable-next-line no-shadow
existsSync, readFileSync) {
    var _a;
    if (typeof extendedConfigValue === "string" &&
        extendedConfigValue.indexOf(".json") === -1) {
        extendedConfigValue += ".json";
    }
    const currentDir = path.dirname(configFilePath);
    let extendedConfigPath = path.join(currentDir, extendedConfigValue);
    if (extendedConfigValue.indexOf("/") !== -1 &&
        extendedConfigValue.indexOf(".") !== -1 &&
        !existsSync(extendedConfigPath)) {
        extendedConfigPath = path.join(currentDir, "node_modules", extendedConfigValue);
    }
    const config = loadTsconfig(extendedConfigPath, existsSync, readFileSync) || {};
    // baseUrl should be interpreted as relative to extendedConfigPath,
    // but we need to update it so it is relative to the original tsconfig being loaded
    if ((_a = config.compilerOptions) === null || _a === void 0 ? void 0 : _a.baseUrl) {
        const extendsDir = path.dirname(extendedConfigValue);
        config.compilerOptions.baseUrl = path.join(extendsDir, config.compilerOptions.baseUrl);
    }
    return config;
}
function mergeTsconfigs(base, config) {
    base = base || {};
    config = config || {};
    return Object.assign(Object.assign(Object.assign({}, base), config), { compilerOptions: Object.assign(Object.assign({}, base.compilerOptions), config.compilerOptions) });
}
