// @babel/core/src/config/files/index-browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPreset = exports.loadPlugin = exports.resolvePreset = exports.resolvePlugin = exports.ROOT_CONFIG_FILENAMES = exports.resolveShowConfigPath = exports.loadConfig = exports.findRootConfig = exports.findRelativeConfig = exports.findPackageData = exports.findConfigUpwards = void 0;
function findConfigUpwards(
// eslint-disable-next-line @typescript-eslint/no-unused-vars
rootDir) {
    return null;
}
exports.findConfigUpwards = findConfigUpwards;
// eslint-disable-next-line require-yield
function* findPackageData(filepath) {
    return {
        filepath,
        directories: [],
        pkg: null,
        isPackage: false,
    };
}
exports.findPackageData = findPackageData;
// eslint-disable-next-line require-yield
function* findRelativeConfig(
// eslint-disable-next-line @typescript-eslint/no-unused-vars
pkgData, 
// eslint-disable-next-line @typescript-eslint/no-unused-vars
envName, 
// eslint-disable-next-line @typescript-eslint/no-unused-vars
caller) {
    return { config: null, ignore: null };
}
exports.findRelativeConfig = findRelativeConfig;
// eslint-disable-next-line require-yield
function* findRootConfig(
// eslint-disable-next-line @typescript-eslint/no-unused-vars
dirname, 
// eslint-disable-next-line @typescript-eslint/no-unused-vars
envName, 
// eslint-disable-next-line @typescript-eslint/no-unused-vars
caller) {
    return null;
}
exports.findRootConfig = findRootConfig;
// eslint-disable-next-line require-yield
function* loadConfig(name, dirname, 
// eslint-disable-next-line @typescript-eslint/no-unused-vars
envName, 
// eslint-disable-next-line @typescript-eslint/no-unused-vars
caller) {
    throw new Error(`Cannot load ${name} relative to ${dirname} in a browser`);
}
exports.loadConfig = loadConfig;
// eslint-disable-next-line require-yield
function* resolveShowConfigPath(
// eslint-disable-next-line @typescript-eslint/no-unused-vars
dirname) {
    return null;
}
exports.resolveShowConfigPath = resolveShowConfigPath;
exports.ROOT_CONFIG_FILENAMES = [];
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function resolvePlugin(name, dirname) {
    return null;
}
exports.resolvePlugin = resolvePlugin;
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function resolvePreset(name, dirname) {
    return null;
}
exports.resolvePreset = resolvePreset;
function loadPlugin(name, dirname) {
    throw new Error(`Cannot load plugin ${name} relative to ${dirname} in a browser`);
}
exports.loadPlugin = loadPlugin;
function loadPreset(name, dirname) {
    throw new Error(`Cannot load preset ${name} relative to ${dirname} in a browser`);
}
exports.loadPreset = loadPreset;
